File: evaluation/results/scenario_3/pymc/llama3.1:8b/run_03/extracted_code.py
import pandas as pd
import numpy as np
from pymc import Model, StepMethod, SampleFromPrior, NUTS
import pymc as pm
import matplotlib.pyplot as plt

# Load data
data = pd.read_csv('observations.csv')

# Define model parameters
n_cases = len(data['cases'])
n_hospitalisations = len(data['hospitalisations'])
n_deaths = len(data['deaths'])

# Delay and ascertainment rates for each stream
delay_cases = 0.5  # Example delay, adjust as needed
delay_hospitalisations = 3
delay_deaths = 7

ascertainment_rates = [1/6, 1/12, 1/20]  # Adjust these as needed to reflect proportions of infections observed

# Prior distribution for Rt
rtprior = pm.Uniform('Rt', lower=0.5, upper=2)

# Infection dynamics model using renewal equation
with Model() as model:
    Rt = rtprior

    cases_observed = pm.Poisson('cases_observed', mu=Rt*pm.math.sum([cases[(delay_cases + s) % n_cases] for s in range(n_cases)], axis=0))
    hospitalisations_observed = pm.Poisson('hospitalisations_observed', mu=Rt*pm.math.sum([hospitalisations[(delay_hospitalisations + s) % n_hospitalisations] for s in range(n_hospitalisations)], axis=0))
    deaths_observed = pm.Poisson('deaths_observed', mu=Rt*pm.math.sum([deaths[(delay_deaths + s) % n_deaths] for s in range(n_deaths)], axis=0))

    # Overdispersion to account for greater variance than Poisson
    cases_observed_dist = pm.Deterministic('cases_observed_dist', cases_observed * (1+pm.math.sqrt(pm.math.log(n_cases))))
    hospitalisations_observed_dist = pm.Deterministic('hospitalisations_observed_dist', hospitalisations_observed * (1+pm.math.sqrt(pm.math.log(n_hospitalisations))))
    deaths_observed_dist = pm.Deterministic('deaths_observed_dist', deaths_observed * (1+pm.math.sqrt(pm.math.log(n_deaths))))

# Initialising the seed
data['cases'][0] = 100  # Assign a dummy value to represent initial infections

# Model fitting and MCMC sampling
with model:
    trace = pm.sample(2000, tune=1000)

# Extract Rt estimates
rtestimates = trace['Rt']

# Plotting the results
plt.plot(rtestimates)
plt.title('Time-Varying Reproduction Number (Rt)')
plt.xlabel('Time')
plt.ylabel('Reproduction Number')
plt.show()

print(pm.summary(trace))

File: evaluation/results/scenario_3/turing/llama3.1:8b/run_03/extracted_code.jl
using Turing, DataFrames, StatsPlots

# Load data
data = readcsv("observations.csv")
df = DataFrame(data)

# Define model
@model function Rt_model(observations)
    R_t ~ TruncatedNormal(0.5, 1.0, lower=0) # Shared reproduction number

    # Model each stream separately to account for different delays and ascertainment rates
    for (i, row) in enumerate(eachrow(observations))
        date = Dates.parse(row.date)

        # Define observation-specific parameters
        cases_delay = Dates.value(Dates.day(Date("2022-02-01"), date))  # Assume hospitalisations and deaths are reported with a fixed delay of approximately 14 days after infection onset for most data sets. Adjust according to the actual delays in your dataset.
        hospitalisations_delay = cases_delay + 14
        deaths_delay = cases_delay + 28

        cases_observed = row.cases
        hospitalisations_observed = row.hospitalisations
        deaths_observed = row.deaths

        # Model the number of infections given the past infections and reproduction number
        logit_cases = R_t * sum(@. observations[i-s, :cases] / (1 + cases_delay))  # Assume ascertainment rate is 1 for simplicity in this example; replace with actual rates if available.
        logit_hospitalisations = R_t * sum(@. observations[i-s, :hospitalisations] / (1 + hospitalisations_delay))
        logit_deaths = R_t * sum(@. observations[i-s, :deaths] / (1 + deaths_delay))

        # Fit overdispersed Poisson distribution for each stream
        cases ~ InverseGamma(α=1, β=1) .* (exp.(logit_cases)) ./ (1 .+ exp.(logit_cases))
        hospitalisations ~ InverseGamma(α=1, β=1) .* (exp.(logit_hospitalisations)) ./ (1 .+ exp.(logit_hospitalisations))
        deaths ~ InverseGamma(α=1, β=1) .* (exp.(logit_deaths)) ./ (1 .+ exp.(logit_deaths))

    end
end

# Define a function to fit the model and extract Rt estimates
function fit_model(data)
    observations = [nrow(df)-i for i in eachindex(eachcol(df))]
    chain = sample(Rt_model(observations), MCMCThreads(), 1000, burnin=500, thinning=5)

    # Extract Rt estimates and stream-specific parameters
    rt_estimates = mean(chain[:R_t])
    return rt_estimates, chain
end

# Fit the model
rt_estimate, _ = fit_model(df)
println("Estimated reproduction number (Rt): ", rt_estimate)

# Plot results
plot([Dates.parse.(df.date) for i in eachindex(eachrow(df))], rt_estimate,
     label="Estimated Rt", xlabel="Time", ylabel="Reproduction Number",
     title="Estimated Time-Varying Reproduction Number")

